### DISCUSSION TITLE
Rack 3, json_parser and Rack::MethodOverride - JSON not copied to params

### DISCUSSION THREAD

**Discussion by jcsjcs** [jeremyevans/roda#291] [General]
After upgrading to Rack 3, the json_parser plugin is no longer unpacking JSON body into params if rack's method override is used.

Here is a small example app showing the behaviour (comment/uncomment `use Rack::MethodOverride` to see the effect):
```ruby
class App < Roda
  # use Rack::MethodOverride

  plugin :symbolized_params
  plugin :json_parser

  route do |r|
    r.on 'json_in' do
      p params
      { params_received: params }.to_json
    end
  end
end
```
```sh
# Without method override:
curl --data '{"some":"json"}' --header 'Content-Type: application/json' http://localhost:9299/json_in
{"params_received":{"some":"json"}

# With method override:
curl --data '{"some":"json"}' --header 'Content-Type: application/json' http://localhost:9299/json_in
{"params_received":{}}
```
It looks like this Rack pull request triggered the change in behaviour: https://github.com/rack/rack/pull/1795

I would like to continue using `json_parser` and `all_verbs` with Rack::MethodOverride together.
Is there a way to work around this?

**Comment by jeremyevans** [comment]
I think you should submit a pull request to rack to update `Rack::MethodOverride#method_override_param` to do `req.POST[METHOD_OVERRIDE_PARAM_KEY] if req.form_data? || req.parseable_data?`.  That way it won't try to check for the method override in cases where it definitely won't succeed.  That's is something I would approve.

**Reply by jcsjcs** [reply]
Thank you, Jeremy. This works perfectly in my testing and I have opened a pull request: https://github.com/rack/rack/pull/1980
